server/leafnode.go

```diff
@@ -679,9 +679,18 @@ func (s *Server) createLeafNode(conn net.Conn, remote *leafNodeCfg) *client {
 
 			// Force handshake
 			c.mu.Unlock()
-			if err := conn.Handshake(); err != nil {
+			if err = conn.Handshake(); err != nil {
 				c.Errorf("TLS handshake error: %v", err)
 				c.closeConnection(TLSHandshakeError)
+				// If we overrode and used the saved tlsName but that failed
+				// we will clear that here. This is for the case that another server
+				// does not have the same tlsName, maybe only IPs.
+				// https://github.com/nats-io/nats-server/issues/1256
+				if host == c.leaf.remote.tlsName {
+					c.mu.Lock()
```

Comment:
This is a connection failure, that does not happen millions of time a second, so I would have moved the locking above the check, so that check does not access `c.leaf.remote` without locking...

Reply:
fixed.

Reply:
Fair..

Reply:
fixed